src/test/scala/org/apache/spark/sql/pulsar/PulsarAdmissionControlSuite.scala

```diff
@@ -0,0 +1,225 @@
+    sendMessages(topic, Array("-1"))
+    require(getLatestOffsets(Set(topic)).size === 1)
+    
+    val pulsar = spark.readStream
+      .format("pulsar")
+      .option(TopicSingle, topic)
+      .option(ServiceUrlOptionKey, serviceUrl)
+      .option(AdminUrlOptionKey, adminUrl)
+      .option(MaxBytesPerTrigger, 150)
+      .load()
+      .selectExpr("CAST(__key AS STRING)", "CAST(value AS STRING)")
+      .as[(String, String)]
+
+    val mapped = pulsar.map(kv => kv._2.toInt + 1)
+
+    // Each Int adds 49 bytes to message size, so we expect 3 Ints in each message
+    testStream(mapped)(
+      StartStream(trigger = ProcessingTime(1000)),
+      makeSureGetOffsetCalled,
+      AddPulsarData(Set(topic), 1, 2, 3),
+      CheckLastBatch(2, 3, 4),
+      AddPulsarData(Set(topic), 4, 5, 6, 7, 8, 9),
+      CheckLastBatch(8, 9, 10),
+      AssertOnQuery { query =>
+        val recordsRead = query.recentProgress.map(_.numInputRows).sum
+        recordsRead == 9
+      }
+    )
+  }
+
+  test("Only admit first entry of ledger") {
+    val topic = newTopic()
+    val messageIds = sendMessages(topic, Array("1", "2", "3"))
+    val firstMid = messageIds.head._2
+    val firstLedger = getLedgerId(firstMid)
+    val firstEntry = getEntryId(firstMid)
+    require(getLatestOffsets(Set(topic)).size === 1)
+    val admissionControlHelper = new PulsarAdmissionControlHelper(adminUrl)
+    val offset = admissionControlHelper.latestOffsetForTopicPartition(topic, MessageId.earliest, 1)
+    assert(getLedgerId(offset) == firstLedger && getEntryId(offset) == firstEntry)
+
+  }
+
+  test("Admit entry in the middle of the ledger") {
+    val topic = newTopic()
+    val messageIds = sendMessages(topic, Array("1", "2", "3"))
+    val firstMid = messageIds.head._2
+    val secondMid = messageIds.apply(1)._2
+    require(getLatestOffsets(Set(topic)).size === 1)
+    val admissionControlHelper = new PulsarAdmissionControlHelper(adminUrl)
+    val offset = admissionControlHelper.latestOffsetForTopicPartition(topic, firstMid, 1)
+    assert(getLedgerId(offset) == getLedgerId(secondMid) && getEntryId(offset) == getEntryId(secondMid))
+
+  }
+
+  test("Admission Control for multiple topics") {
+    val topic1 = newTopic()
+    val topic2 = newTopic()
+    
+    val pulsar = spark.readStream
+      .format("pulsar")
+      .option(TopicMulti, s"$topic1,$topic2")
+      .option(ServiceUrlOptionKey, serviceUrl)
+      .option(AdminUrlOptionKey, adminUrl)
+      .option(MaxBytesPerTrigger, 300)
+      .load()
+      .selectExpr("CAST(__key AS STRING)", "CAST(value AS STRING)")
+      .as[(String, String)]
+
+    val mapped = pulsar.map(kv => kv._2.toInt + 1)
+
+    // Each Int adds 49 bytes to message size, so we expect 3 Ints in each message
+    testStream(mapped)(
+      StartStream(trigger = ProcessingTime(1000)),
+      makeSureGetOffsetCalled,
+      AddPulsarData(Set(topic1), 1, 2, 3),
+      CheckLastBatch(2, 3, 4),
+      AddPulsarData(Set(topic2), 4, 5, 6, 7, 8, 9),
```

Comment:
Can we send data to multiple topics at the same time?

Reply:
Added test case